source/slang/slang-check-decl.cpp

```diff
@@ -10178,20 +10183,22 @@ namespace Slang
 
     CapabilitySet getStatementCapabilityUsage(SemanticsVisitor* visitor, Stmt* stmt);
 
-    template<typename ProcessFunc>
+    template<typename ProcessFunc, typename RefParentDiagnosticFunc>
     struct CapabilityDeclReferenceVisitor
-        : public SemanticsDeclReferenceVisitor<CapabilityDeclReferenceVisitor<ProcessFunc>>
+        : public SemanticsDeclReferenceVisitor<CapabilityDeclReferenceVisitor<ProcessFunc, RefParentDiagnosticFunc>>
     {
-        typedef SemanticsDeclReferenceVisitor<CapabilityDeclReferenceVisitor<ProcessFunc>> Base;
+        typedef SemanticsDeclReferenceVisitor<CapabilityDeclReferenceVisitor<ProcessFunc, RefParentDiagnosticFunc>> Base;
 
         const ProcessFunc& handleReferenceFunc;
+        const RefParentDiagnosticFunc& handleReferenceParentDiagnosticFunc;
         RequireCapabilityAttribute* maybeRequireCapability;
         SemanticsContext& outerContext;
-        CapabilityDeclReferenceVisitor(const ProcessFunc& processFunc, RequireCapabilityAttribute* maybeRequireCapability, SemanticsContext& outer)
+        CapabilityDeclReferenceVisitor(const ProcessFunc& processFunc, const RefParentDiagnosticFunc& refParentDiagnosticFunc, RequireCapabilityAttribute* maybeRequireCapability, SemanticsContext& outer)
             : handleReferenceFunc(processFunc)
+            , handleReferenceParentDiagnosticFunc(refParentDiagnosticFunc)
```

Comment:
I think the name is too long and it became harder to read.
I like to suggest to rename the existing variable "handleReferenceFunc" to just "processFunc" and the new one to "parentDiagnosticFunc".
The keywords, "Reference" and "handle", don't carry any meaning.

I also like to note that functors in C++ are normally copiable objects and passed around as a copy not as a reference.
The uses of `&` here seem unnecessary and it can be a potential problem later if the functor is originated from a stack memory and it is de-referenced after the stack became invalid.

Reply:
I agree with renaming variables. 

> I also like to note that functors in C++ are normally copiable objects and passed around as a copy not as a reference.

A copy can be used here, I will switch the types around.
I think the reason we copy by reference here is because lambdas do not decay into a simple function pointer in most cases, lambdas can explode in size very quickly depending on the parameters captured.